js/core/template_manager.js

```diff
@@ -0,0 +1,221 @@
+import $ from './renderer';
+import { isDefined, isFunction, isRenderer } from './utils/type';
+import { noop } from './utils/common';
+import { extend } from './utils/extend';
+import { Error, log } from './errors';
+import { getElementOptions } from './utils/dom';
+import { TemplateBase } from './templates/template_base';
+import { FunctionTemplate } from './templates/function_template';
+import { EmptyTemplate } from './templates/empty_template';
+import { ChildDefaultTemplate } from './templates/child_default_template';
+import { camelize } from './utils/inflector';
+import {
+    findTemplateByDevice, addOneRenderedCall, templateKey,
+    getNormalizedTemplateArgs, validateTemplateSource,
+    defaultCreateElement,
+} from './utils/template_manager';
+
+const TEXT_NODE = 3;
+const ANONYMOUS_TEMPLATE_NAME = 'template';
+const TEMPLATE_SELECTOR = '[data-options*="dxTemplate"]';
+const TEMPLATE_WRAPPER_CLASS = 'dx-template-wrapper';
+const DEPRECATED_WIDGET_NAMES = new Set(['button', 'tabs', 'dropDownMenu']);
+const DX_POLYMORPH_WIDGET_TEMPLATE = new FunctionTemplate(({ model, parent }) => {
+    let widgetName = model.widget;
+    if(!widgetName) return $();
+
+    const widgetElement = $('<div>');
+    const widgetOptions = model.options || {};
+
+    if(DEPRECATED_WIDGET_NAMES.has(widgetName)) {
+        const deprecatedName = widgetName;
+        widgetName = camelize('dx-' + widgetName);
+        log('W0001', 'dxToolbar - "widget" item field', deprecatedName, '16.1', 'Use: ' + widgetName + 'instead');
+    }
+
+    if(parent) {
+        parent._createComponent(widgetElement, widgetName, widgetOptions);
+    } else {
+        widgetElement[widgetName](widgetOptions);
+    }
+
+    return widgetElement;
+});
+
+export default class TemplateManager {
+    constructor(createElement, anonymousTemplateName) {
+        this._tempTemplates = [];
+        this._defaultTemplates = {};
+        this._anonymousTemplateName = anonymousTemplateName || ANONYMOUS_TEMPLATE_NAME;
+
+        this._createElement = createElement || defaultCreateElement;
+        this._createTemplateIfNeeded = this._createTemplateIfNeeded.bind(this);
+    }
+
+    static get defaultOptions() {
```

Comment:
This seems to be a method, as it returns a new object on each call. `buildDefaultOptions` or smth like.
@vconst, @MaximKudriavtsev I see your discussion below on the topic `getter vs function` and this one seems to me to be a method